### PR TITLE
Added the Singularity Functions module.

### PR DESCRIPTION
**Pull request by sampadsaha5**
- [x] SingularityFunction Class and its tests.
- [x] `fdiff()` and its tests
- [x] `rewrite(Piecewise)` and its test
- [x] Pretty Printing.
- [x] Latex Printing.
- [x] Rewrite as Heaviside/DiracDelta.

Ping @moorepants @leosartaj

### CODE REVIEW

**Comment by sampadsaha5** [comment]
I tried adding : 
```py
    def __repr__(self):
        str_sol = 'SingularityFunction(%s, %s, %s)' % (sstr(self.args[0]), sstr(self.args[1]), sstr(self.args[2]))
        return str_sol
```
and i have added heaviside and diracdelta under eval method but still i have this output:
```py
>>> SingularityFunction(x, 5, 4)
(x-5)**4*Heaviside(x-5)             
```
But the desired one is:
```py
>>> SingularityFunction(x, 5, 4)
SingularityFunction(x, 5, 4)             
```

**Comment by sampadsaha5** [comment]
Added the functionality of pretty printing:
```py
In [5]: SingularityFunction(x,y,n)
       n
<x - y> 

In [6]: SingularityFunction(x,4,n)
       n
<x - 4> 

In [7]: (SingularityFunction(x,-5,n)
       n
<x + 5> 

In [8]: SingularityFunction(x,0,n)
   n
<x> 

```
Ping @moorepants @leosartaj

**Comment by sampadsaha5** [comment]
Ping @moorepants @leosartaj.

File: sympy/functions/special/tests/test_singularity_functions.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    assert SingularityFunction(5, 3, 2) == 4
+    assert SingularityFunction(3, 5, 1) == 0
+    assert SingularityFunction(3, 3, 0) == 1
+    assert SingularityFunction(4, 4, -1) == oo
+    assert SingularityFunction(4, 2, -1) == 0
+    assert SingularityFunction(4, 7, -1) == 0
+    assert SingularityFunction(5, 6, -2) == 0
+    assert SingularityFunction(4, 2, -2) == 0
+    assert SingularityFunction(4, 4, -2) == oo
+    assert SingularityFunction(6.1, 4, 5) == 40.84100999999997
```
Comment (moorepants) [inline sympy/functions/special/tests/test_singularity_functions.py]: Comparing floats should be done using sympy Float type or you need to use proper floating point comparison.

Nice reference to read: https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html

File: sympy/functions/special/tests/test_singularity_functions.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    assert SingularityFunction(6.1, pi, 2) == (-pi + 6.1)**2
+    assert SingularityFunction(x, a, nan) == nan
+    assert SingularityFunction(x, nan, 1) == nan
+    assert SingularityFunction(nan, a, n) == nan
+    assert conjugate(SingularityFunction(x, -10, 5)) == SingularityFunction(x, -10, 5)
+    assert adjoint(SingularityFunction(x, -10, 5)) == SingularityFunction(x, -10, 5)
+    assert transpose(SingularityFunction(x, -10, 5)) == SingularityFunction(x, -10, 5)
```
Comment (moorepants) [inline sympy/functions/special/tests/test_singularity_functions.py]: Find out if these three tests are correct and remove if not or right correct tests.

**Comment by sampadsaha5** [comment]
I think rewriting as heaviside and diracdelta would be in the same function.
the function would be like:
```py
    def _eval_rewrite_as_HeavisideDiracDelta(self, *args):
        x = self.args[0]
        a = self.args[1]
        n = self.args[2]

        if n == -2:
            return diff(Heaviside(x - a), x, 2)
        if n == -1:
            return diff(Heaviside(x - a), x, 1)
        if n >= 0:
            return (x - a)**n*Heaviside(x - a)

```
The API would be somewhat:-
```py
In [1]: expr = SingularityFunction(x, 4, 5) + SingularityFunction(x, -3, -1) - SingularityFunction(x, 0, -2)

In [2]: expr.rewrite(HeavisideDiracDelta)
Out[2]: (x - 4)**5*Heaviside(x - 4) + DiracDelta(x + 3) - DiracDelta(x, 1)

```
Any suggestions are welcome.

**Comment by leosartaj** [comment]
Does this even work(The function name appears weird)? You tried implementing it?

**Comment by sampadsaha5** [comment]
yeah, I just checked that. It is not working with the name "HeavisideDiracDelta". But it is completely fine if "HeavisideDiracDelta" is replaced by "Heaviside" or may be "DiracDelta".

**Comment by sampadsaha5** [comment]
> I think rewriting as heaviside and diracdelta would be in the same function.

The reason for this is:
when `n == -2` or `n == -1 ` part gets executed then it results into a `DiracDelta` expression. So to get a Heaviside expression for `n == -2` or `n == -1 ` we have to use:
 `diff(Heaviside(x - a), x, 2, evaluate = False)` and
`diff(Heaviside(x - a), x, 1, evaluate = False)` respectively.

But if we integrate an unevaluated instance of derivative then it raises an error.  The same happens with diracdelta also.

So I feel that writing a method which would return the heaviside expressions exclusively or the DiracDelta expression exclusively would be of no use. It would be better to rewrite as heaviside and diracdelta both in a same method. We just have to think of a "good name".

File: sympy/functions/special/tests/test_singularity_functions.py (hunk 2 above)
Comment (sampadsaha5) [inline sympy/functions/special/tests/test_singularity_functions.py]: Sure.

**Comment by moorepants** [comment]
I think you simply need to do this:

```python
def rewrite_Heaviside(self):
    # bunch of code to implement

def rewrite_DiracDelta(self):
    return self.rewrite_Heaviside
```

**Comment by sampadsaha5** [comment]
Ah.. I got you there.
This way we can use any one of the methods.
`rewrite(Heaviside)`  or `rewrite(DiracDelta)`. Both would return an expression containing DiracDelta and Heaviside both.

Great idea

**Comment by sampadsaha5** [comment]
Ping @moorepants @leosartaj

**Comment by sampadsaha5** [comment]
This also works fine:
```py
    def _eval_rewrite_as_HeavisideDiracDelta(self, *args):
        x = self.args[0]
        a = self.args[1]
        n = self.args[2]

        if n == -2:
            return diff(Heaviside(x - a), x, 2)
        if n == -1:
            return diff(Heaviside(x - a), x, 1)
        if n >= 0:
            return (x - a)**n*Heaviside(x - a)
```
The API would be somewhat:-
```py
In [1]: expr = SingularityFunction(x, 4, 5) + SingularityFunction(x, -3, -1) - SingularityFunction(x, 0, -2)

In [2]: expr.rewrite('HeavisideDiracDelta')
Out[2]: (x - 4)**5*Heaviside(x - 4) + DiracDelta(x + 3) - DiracDelta(x, 1)
```

**Comment by sampadsaha5** [comment]
Can we convert `a*b*c*f(DiracDelta(h(x)))*d*e` to a SingularityFunction ?

File: sympy/functions/special/tests/test_singularity_functions.py (hunk 2 above)
Comment (sampadsaha5) [inline sympy/functions/special/tests/test_singularity_functions.py]: I found out that there are no such things defined which are known as either conjugate or transpose or adjoint of Singularity Functions.

**Comment by sampadsaha5** [comment]
Suppose we have an expression `(x -5)**3*Heaviside(x - 5)` or may be a bit more complex one like `f(x - a)*Heaviside(x- a)`, so here how can we rewrite it as Singularity Functions. I have tried one method which returns as `(x - 5)**3*SingularityFunction(x, 5, 0))` but I want it as `SingularityFunction(x, 5, 3)` if `f(x - a)` is of that form.

**Comment by sampadsaha5** [comment]
p = exp(-(x**2 + y**2))/pi
p*DiracDelta(x - 10*y)

**Comment by sampadsaha5** [comment]
This PR is now ready for review.
Ping @moorepants @leosartaj .

**Comment by sampadsaha5** [comment]
Ping @moorepants @leosartaj . I have added the suggested changes .

File: sympy/functions/special/singularity_functions.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+
+class SingularityFunction(Function):
+    r"""
+    The Singularity functions are a class of discontinuous functions. They take a
+    variable, an offset and an exponent as arguments. These functions are
+    represented using Macaulay brackets as :
+
+    .. math::
+        SingularityFunction(x, a, n) := {\langle x - a \rangle}^n
```
Comment (moorepants) [inline sympy/functions/special/singularity_functions.py]: This is not easily readable at the command line. I suggest simply using:

`SingularityFunction(x, a, n) := <x - a>^n`

File: sympy/functions/special/singularity_functions.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    Returns ``Derivative(DiracDelta(x - a), x, n + 1)`` if ``n < 0``
+    Returns ``(x - a)**n*Heaviside(x - a)`` if ``n >= 0``
```
Comment (moorepants) [inline sympy/functions/special/singularity_functions.py]: What does this mean? How does a class return something?

Comment (moorepants) [inline sympy/functions/special/singularity_functions.py]: "The singularity function will automatically evaluate to X and Y if n < 0 and n >= 0."

File: sympy/functions/special/singularity_functions.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    ========
+
+    >>> from sympy import SingularityFunction, diff, Piecewise, DiracDelta, Heaviside, Symbol
+    >>> from sympy.abc import x, a, n
+    >>> SingularityFunction(x, a, n)
+    SingularityFunction(x, a, n)
+    >>> y = Symbol('y', positive=True)
+    >>> n = Symbol('n', nonnegative=True)
+    >>> SingularityFunction(y, -10, n)
+    (y + 10)**n
```
Comment (moorepants) [inline sympy/functions/special/singularity_functions.py]: Can you add a test or example that shows this works:

```pycon
>>> expr = SingularityFunction(x, a, n)
>>> y = Symbol('y', positive=True)
>>> n = Symbol('n', nonnegative=True)
>>> expr.subs({x: y, a: -10, n: n})
(y + 10)**n
```

File: sympy/functions/special/tests/test_singularity_functions.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    assert SingularityFunction(5, 3, 2) == 4
+    assert SingularityFunction(3, 5, 1) == 0
+    assert SingularityFunction(3, 3, 0) == 1
+    assert SingularityFunction(4, 4, -1) == oo
+    assert SingularityFunction(4, 2, -1) == 0
+    assert SingularityFunction(4, 7, -1) == 0
+    assert SingularityFunction(5, 6, -2) == 0
+    assert SingularityFunction(4, 2, -2) == 0
+    assert SingularityFunction(4, 4, -2) == oo
+    assert float(SingularityFunction(6.1, 4, 5)) == 40.84100999999997
```
Comment (moorepants) [inline sympy/functions/special/tests/test_singularity_functions.py]: This is not good practice for comparing floats. You need to substract and compare as based on a tolerance. See http://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.testing.assert_allclose.html for an example function for doing this. There may be a convenience float comparision function in sympy for this. Or use the sympy Float object which is arbitrary precision.

**Comment by sampadsaha5** [comment]
Ping @leosartaj @moorepants .
I have added all the suggested modifications.

File: sympy/functions/special/tests/test_singularity_functions.py [hunk 7]
```diff
@@ -46,7 +46,7 @@ def test_eval():
     assert SingularityFunction(5, 6, -2) == 0
     assert SingularityFunction(4, 2, -2) == 0
     assert SingularityFunction(4, 4, -2) == oo
-    assert float(SingularityFunction(6.1, 4, 5)) == 40.84100999999997
+    assert (SingularityFunction(6.1, 4, 5)).evalf(5) == 40.84100
```
Comment (moorepants) [inline sympy/functions/special/tests/test_singularity_functions.py]: This is still not correct. The right side is a python float where as the left side is a SymPy Float with 5 digits. You could change the right side to `Float('40.84100')` to fix it. It is incorrect to compare two Python floats using `==`.

Comment (moorepants) [inline sympy/functions/special/tests/test_singularity_functions.py]: Please read this: http://docs.sympy.org/dev/gotchas.html to see how sympy floats, python floats, and sympy evaluation work.

File: sympy/functions/special/singularity_functions.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+class SingularityFunction(Function):
+    r"""
+    The Singularity functions are a class of discontinuous functions. They take a
+    variable, an offset and an exponent as arguments. These functions are
+    represented using Macaulay brackets as :
+
+    SingularityFunction(x, a, n) := <x - a>^n
+
+    The singularity function will automatically evaluate to
+    ``Derivative(DiracDelta(x - a), x, n + 1)`` if ``n < 0``
```
Comment (jksuom) [inline sympy/functions/special/singularity_functions.py]: I think the count argument of `Derivative` should be nonnegative. So maybe it should be `-n - 1`.

Comment (sampadsaha5) [inline sympy/functions/special/singularity_functions.py]: Yah you are right. Thanks @jksuom.

**Comment by sampadsaha5** [comment]
Ping @leosartaj @moorepants .

**Comment by moorepants** [comment]
@sampadsaha5 I think you need to merge in master and then this will be ready to be merged. +1

**Comment by sampadsaha5** [comment]
Thanks @moorepants .

**Comment by sampadsaha5** [comment]
After running a quality test I am having an error like this : 
```py
ahappyidiot@Ahappyidiot:~/Desktop/Gsoc/sympy$ bin/test quality
================================================================= test process starts =================================================================
executable:         /home/ahappyidiot/anaconda2/bin/python  (2.7.11-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        54461817
hash randomization: on (PYTHONHASHSEED=2060343345)

sympy/utilities/tests/test_code_quality.py[6] F.....                                                                                             [FAIL]

_______________________________________________________________________________________________________________________________________________________
________________________________________________ sympy/utilities/tests/test_code_quality.py:test_files ________________________________________________
  File "/home/ahappyidiot/Desktop/Gsoc/sympy/sympy/utilities/tests/test_code_quality.py", line 227, in test_files
    check_directory_tree(BIN_PATH, test, set(["~", ".pyc", ".sh"]), "*")
  File "/home/ahappyidiot/Desktop/Gsoc/sympy/sympy/utilities/tests/test_code_quality.py", line 107, in check_directory_tree
    check_files(glob(join(root, pattern)), file_check, exclusions)
  File "/home/ahappyidiot/Desktop/Gsoc/sympy/sympy/utilities/tests/test_code_quality.py", line 123, in check_files
    file_check(fname)
  File "/home/ahappyidiot/Desktop/Gsoc/sympy/sympy/utilities/tests/test_code_quality.py", line 146, in test
    test_this_file(fname, test_file)
  File "/home/ahappyidiot/Desktop/Gsoc/sympy/sympy/utilities/tests/test_code_quality.py", line 162, in test_this_file
    assert False, message_test_suite_def % (fname, idx + 1)
AssertionError: Function should start with 'test_' or '_': /home/ahappyidiot/Desktop/Gsoc/sympy/bin/generate_test_list.py, line 57
```

**Comment by jksuom** [comment]
I get the same error in current master. It also refers to `bin/generate_test_list.py`.

**Comment by sampadsaha5** [comment]
Yah it is there in master branch.

**Comment by sampadsaha5** [comment]
@jksuom I think an issue should be reported.

**Comment by jksuom** [comment]
Yes. And it is odd that the error does not show in travis tests.
Ping @asmeurer

**Comment by moorepants** [comment]
+1 to merge this once tests pass.

**Pull request merged**